bot/src/export.rs

```diff
@@ -0,0 +1,638 @@
+        None,
+    )?;
+
+    info!("Got {} participants", participants.len());
+
+    for (_, participant) in participants {
+        let rows = transaction.query(
+            "SELECT id, state from \"ValidatorKeyPair\" \
+            WHERE mainnet_beta_pk=$1 AND \
+            testnet_pk=$2",
+            &[
+                &participant.mainnet_identity.to_string(),
+                &participant.testnet_identity.to_string()
+            ])?;
+
+        // this strange incantation get the string without surrounding quotation marks
+        let previous_state = &serde_json::to_value(&participant.state)?.as_str().unwrap().to_string();
+
+        if rows.len() == 0 {
+            transaction.execute(
+                "INSERT INTO \"ValidatorKeyPair\"\
+                                    (mainnet_beta_pk, testnet_pk, state) \
+                                    VALUES ($1, $2, $3)",
+                &[
+                    &participant.mainnet_identity.to_string(),
+                    &participant.testnet_identity.to_string(),
+                    previous_state
+                ])?;
+        } else if rows.len() == 1 {
+            let row = rows.first().unwrap();
+
+            let current_state: String = row.get("state");
+
+            if &current_state != previous_state {
+                let id: i32 = row.get("id");
+
+                transaction.execute(
+                    "UPDATE \"ValidatorKeyPair\"\
+                SET state=$1 \
+                WHERE id=$2",
+                    &[
+                        &previous_state,
+                        &id
+                    ])?;
+            }
+        } else {
+            // DB constraints _should_ make it so this never happens
+            return Err(format!("More than one row in ValidatorKeyPair for main_pk \"{}\", test pk \"{}\"", participant.mainnet_identity, participant.testnet_identity).into());
+        }
+    }
+    Ok(())
+}
+
+/// returns the names for the validators
+fn get_validator_names(rpc_client: &RpcClient) -> BoxResult<HashMap<Pubkey, Map<String, Value>>> {
+    let mut data_map: HashMap<Pubkey, Map<_, _>> = HashMap::new();
+
+    // this value is probably available somewhere
+    let s: &str = "Config1111111111111111111111111111111111111";
+
+    let all_stake_accounts = rpc_client.get_program_accounts(&Pubkey::from_str(s).unwrap())?;
+    for (_, account) in all_stake_accounts {
+        let key_list: ConfigKeys = deserialize(&account.data)?;
+
+        if !key_list.keys.is_empty() {
+            let (validator_pubkey, _) = key_list.keys[1];
+
+            let validator_info_string: String = deserialize(get_config_data(&account.data)?)?;
+            let validator_info: Map<_, _> = serde_json::from_str(&validator_info_string)?;
+            data_map.insert(validator_pubkey, validator_info);
+        }
+    }
+
+    Ok(data_map)
+}
+
+lazy_static! {
+    // TODO: update yml file so that these values can be accessed without using regexes like this
```

Comment:
Ugh yeah we could totally move this all to stronger typing with enums.  The main problem in doing so is re-processing the previous epoch data into the new format, maybe we don't bother though

Reply:
Yeah, for now we decided to not import the historical data.